modules/dr-recover-lost-control-plane-hosts.adoc

```diff
@@ -0,0 +1,278 @@
+// Module included in the following assemblies:
+//
+// * disaster_recovery/scenario-1-infra-recovery.adoc
+
+[id="dr-scenario-1-recover-master-hosts_{context}"]
+= Recovering from lost master hosts
+
+////
+TODO:
+ - Wording on a few steps.
+ - Add a final result
+////
+
+Follow these steps to recover from the loss of one or more master hosts.
+
+.Prerequisites
+
+* Access to the cluster as a user with with the `cluster-admin` role.
+* SSH access to a remaining master host.
+
+.Procedure
+
+. Restore etcd quorum on the remaining master.
++
+[NOTE]
+====
+This step is only necessary if you have had a majority of your masters fail. You can skip this step if you have a majority of your masters still available.
+====
++
+.. Access the remaining master host.
+
+.. Run the `etcd-snapshot-restore.sh` script.
++
+This script takes an optional, but recommended, parameter of an existing snapshot. If no snapshot is specified, it will use the local etcd member's database file. While this data might be more up to date than the backup, the state cannot be guaranteed to be the latest.
++
+----
+[core@ip-10-0-143-125 ~]$ sudo /usr/local/bin/etcd-snapshot-restore.sh <path_to>/snapshot.db
+Creating asset directory ./assets
+Downloading etcdctl binary..
+etcdctl version: 3.3.10
+API version: 3.3
+Backing up /etc/kubernetes/manifests/etcd-member.yaml to ./assets/backup/
+Stopping all static pods..
+..stopping kube-scheduler-pod.yaml
+..stopping kube-controller-manager-pod.yaml
+..stopping kube-apiserver-pod.yaml
+..stopping etcd-member.yaml
+Stopping etcd..
+Waiting for etcd-member to stop
+Stopping kubelet..
+Stopping all containers..
+bd44e4bc942276eb1a6d4b48ecd9f5fe95570f54aa9c6b16939fa2d9b679e1ea
+d88defb9da5ae623592b81619e3690faeb4fa645440e71c029812cb960ff586f
+3920ced20723064a379739c4a586f909497a7b6705a5b3cf367d9b930f23a5f1
+d470f7a2d962c90f3a21bcc021970bde96bc8908f317ec70f1c21720b322c25c
+Backing up etcd data-dir..
+Removing etcd data-dir /var/lib/etcd
+Restoring etcd member etcd-member-ip-10-0-143-125.ec2.internal from snapshot..
+2019-05-15 19:03:34.647589 I | pkg/netutil: resolving etcd-0.clustername.devcluster.openshift.com:2380 to 10.0.143.125:2380
+2019-05-15 19:03:34.883545 I | mvcc: restore compact to 361491
+2019-05-15 19:03:34.915679 I | etcdserver/membership: added member cbe982c74cbb42f [https://etcd-0.clustername.devcluster.openshift.com:2380] to cluster 807ae3bffc8d69ca
+Starting static pods..
+..starting kube-scheduler-pod.yaml
+..starting kube-controller-manager-pod.yaml
+..starting kube-apiserver-pod.yaml
+..starting etcd-member.yaml
+Starting kubelet..
+----
++
+Once the `etcd-snapshot-restore.sh` script completes, your cluster should now have a single member
+etcd cluster, and API services will begin restarting. This might take up to 15 minutes.
++
+In a terminal that has access to the cluster, run the following command to verify that it is ready:
++
```

Comment:
We might want to add here that unless the other etcd members are shutdown it is possible that they will try to connect to the new cluster we just created and spam logs with errors like.

```
2019-05-20 15:33:17.648445 E | rafthttp: request cluster ID mismatch (got 9f5f9f05e4d43b7f want 807ae3bffc8d69ca)
```

This is expected and comes from members from the old cluster 9f5f9f05e4d43b7f trying to connect to 807ae3bffc8d69ca. To resolve make sure the old cluster members are shutdown.

Reply:
Folks see that and they think it is a problem with the new cluster when in fact new cluster is fine but old members still have this peer listed.

Reply:
@hexfusion When you say "unless the other etcd members are shutdown", do you mean the old failed members that we're replacing?

Reply:
Correct, for whatever reason, if one or more of these etcd instances is still running or even crashlooping, users will see logs like above. In which case they should explicitly stop etcd-member static pod on that node to resolve..